---
title: "Upgrading microbiome research in R/Bioc"
author: "Felix G.M. Ernst"
date: "18 12 2020"
output:
    ioslides_presentation:
        widescreen: true
        smaller: true
bibliography: [slides.bib]
biblio-style: apalike
link-citations: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Class structures for microbiome data

```{r phyloseq_class, echo = FALSE, dpi=72, fig.height=6, fig.width=6, fig.align='right', out.extra = 'style="float:right;"'}
BiocPkgTools::plotClassDepData(readRDS("phyloseq_class.rda"))
```

### `phyloseq`

The class mainly used for microbiome analysis

- slots for 
  - count data
  - taxonomic data
  - sequence data
  - tree data

- 8+ years in the making
- similarities to `SummarizedExperiment`

[@R-phyloseq]

## Class structures for microbiome data

### `TreeSummarizedExperiment`

```{r, include=FALSE}
library("TreeSummarizedExperiment")
```

![](tse.svg){width=60%}

- adds tree data to `SingleCellExperiment`
- all the benefits of the `SingleCellExperiment` related packages

[@R-TreeSummarizedExperiment]

## Class structures for microbiome data

require one additional data type to match `phyloseq`: sequencing data

### Two options

1. Create `MicrobiomeExperiment` providing the `refSeq` slot while inheriting 
from `TSE`

or 

2. Add `refSeq` slot to `TSE`

<p>&nbsp;</p>
Currently this is implemented as 1., but the PR for adding it to `TSE` is
also prepared.

## Functions for microbiome workflows {.build}

Specific functions for working with microbiome data need to be 
**added**/**implemented**/**wrapped** to work with `TSE`.

The `mia` package is the playground for this and PRs are very welcome.

### Taxonomic data

- key component of microbiome data
- Instead of implementing a special class, on-the-fly subsetting of `rowData`

```{r, echo=TRUE, message=FALSE}
library(mia)
data("GlobalPatterns")
se <- GlobalPatterns
taxonomyRanks(se)
```

## Functions for microbiome workflows

Specific functions for working with microbiome data need to be 
**added**/**implemented**/**wrapped** to work with `TSE`.

The `mia` package is the playground for this and PRs are very welcome.

### Taxonomic data

- key component of microbiome data
- Instead of implementing a special class, on-the-fly subsetting of `rowData`

```{r, echo=TRUE, message=FALSE}
head(taxonomyRankEmpty(se, "Species"))
all(!taxonomyRankEmpty(se, "Species"))
```

## Functions for microbiome workflows

### Converting taxonomic tabular data to tree data

`TreeSummarizedExperiment` already contains functions for generating data from
hierarchical tabular data. `mia` adds wrapper functions.

```{r taxonomyTree, echo=TRUE, message=FALSE, warning=FALSE}
taxonomyTree(se)
se <- addTaxonomyTree(se)
```

## Functions for microbiome workflows

### Using the taxonomic levels

Data can be aggregated on a selected taxonomic rank.

```{r splitByRanks, echo=TRUE, message=FALSE, warning=FALSE}
x <- agglomerateByRank(se, "Phylum")
dim(se)
dim(x)
```

Internally, constructs a factor to pass to `mia::mergeRows`, which uses
`scater`/`scuttle` functions.

## Functions for microbiome workflows {.build}

### Using the taxonomic levels

`mia::splitByRanks` is a convenience wrapper around `agglomerateByRank`.

Results can be added to the alternative experiment slots inherited from `SCE`.

```{r, echo=TRUE, message=FALSE, warning=FALSE}
altExps(se) <- splitByRanks(se, ranks = taxonomyRanks(se))
altExpNames(se)
```

Analysis can be performed individually on each taxonomic level.

```{r, echo=TRUE, message=FALSE, warning=FALSE}
altExps(se) <-
   lapply(altExps(se),
          function(y){
              rowData(y)$prevalence <-
                  getPrevalence(y, detection = 1/100, sort = FALSE,
                                abund_values = "counts", as_relative = TRUE)
              y
          })
```

## Functions for microbiome workflows

### Using the taxonomic levels

Results from different taxonomic levels can be squashed into a `TSE` using
`mia::unsplitByRanks`.

```{r, echo=TRUE, message=FALSE, warning=FALSE, dpi=72, fig.height=6, fig.width=10}
x <- addTaxonomyTree(unsplitByRanks(se))
x
```


## Functions for microbiome workflows

### Plotting data on a taxonomic tree

Plotting functions split of to `miaViz` package

```{r plotRowTree, echo=TRUE, message=FALSE, warning=FALSE, dpi=72, fig.height=6, fig.width=10}
library(miaViz)
top_phyla <- getTopTaxa(altExp(se,"Phylum"))
plotRowTree(x[rowData(x)$Phylum %in% top_phyla,], edge_colour_by = "Phylum",
            tip_colour_by = "prevalence", node_colour_by = "prevalence")
```

## Summary

**Contributors:**
Tuomas Borman, Domenick J. Braccia, Hector Corrada Bravo, Ruizhu Huang, Leo Lahti, Sudarshan A.Shetty

### Additonal functions already available:

- convert/import functions for `biom`, `phyloseq` and `dada2`
- wrapper for alpha diversity estimates via `vegan` (`breakaway`/`DivNet` not available on `Bioc`, yet?)
- additional reduced dimension methods (JSD, UniFrac)
- in `Bioc-devel` the functions `calculateMDS`/`runMDS` from `scater` accept
  any distance measurement for so that all `vegdist` methods can be used for 
  MDS.
  
<p>&nbsp;</p>
<div style="text-align: center; font-size: 13;"> **Questions? Thoughts? Suggestions?** </div>

## Bibliography

```{r, include=FALSE}
knitr::write_bib(c(.packages(),"phyloseq"),file = "slides.bib")
```
